src/persistent-connections/content.md

```diff
@@ -0,0 +1,465 @@
+
+If you have three Pods, iptables writes these rules:
+
+1. select Pod 1 as the destination with a probability of 0.33
+1. select Pod 2 as the destination with a probability of 0.33
+1. select Pod 3 as the destination with a probability of 0.33
+
+Since this is a probability, there's no guarantee that Pod 2 is selected after Pod 1 as the destination.
+
+> Iptables use the [statistic module](http://ipset.netfilter.org/iptables-extensions.man.html#lbCD) with `random` mode. So the load balancing algorithm is random.
+
+Now that you're familiar with how Services work let's have a look at more exciting scenarios.
+
+## Keeping it alive
+
+With every request started from the front-end to the backend, a new HTTP connection is opened and closed.
+
+If you front-end makes 100 requests per second to the backend, 100 different HTTP connections are opened and closed in that second.
+
+You can improve the latency and save resources if you open a connection and reuse it for any subsequent requests.
+
+The HTTP protocol has a featured called HTTP keep-alive, or HTTP connection reuse that uses a single TCP connection to send and receive multiple HTTP requests and responses.
+
+![Opening and closing connections VS HTTP connection reuse](assets/persistent-1.svg)
+
+It doesn't work out of the box; your server and client should be configured to use it.
+
+But the change is straightforward, and it's available in most languages and frameworks.
+
+Here a few examples on how to implement keep-alive in different languages:
+
+- [Keep-alive in Node.js](https://medium.com/@onufrienkos/keep-alive-connection-on-inter-service-http-requests-3f2de73ffa1)
+- [Keep-alive in Spring boot](https://www.baeldung.com/httpclient-connection-management)
+- [Keep-alive in Python](https://blog.insightdatascience.com/learning-about-the-http-connection-keep-alive-header-7ebe0efa209d)
+- [Keep-alive in .NET](https://docs.microsoft.com/en-us/dotnet/api/system.net.httpwebrequest.keepalive?view=netframework-4.8)
+
+_What happens when you use keep-alive with a Kubernetes Service?_
+
+Let's imagine that front-end and backend support keep-alive.
+
+You have a single instance of the front-end and three replicas for the backend.
+
+The front-end makes the first request to the backend and opens the connection.
+
+The request reaches the Service, and one of the Pod is selected as the destination.
+
+The backend Pod replies and the front-end receives the response.
+
+But instead of closing the HTTP connection, it is kept open for subsequent requests.
+
+_What happens when the front-end issues more requests?_
+
+They are sent to the same Pod.
+
+_Isn't iptables supposed to distribute the traffic?_
+
+It is.
+
+There is a single connection open, and iptables rule were invocated the first time.
+
+One of the three Pods was selected as the destination.
+
+Since all subsequent requests are channelled through the same connection, iptables isn't invoked anymore.
```

Comment:
Since all subsequent requests are channelled through the same **TCP** connection, iptables isn't invoked anymore.

This seems to be the case, as [stated here](https://scalingo.com/blog/iptables#load-balancing):

> Note that the load balancing will only be done during the connection phase of the TCP protocol. Once the connection has been established, the connection will always be routed to the same server.

However, this fact is not obvious and maybe there should be at least a link to a source that confirms this. Because this is the crucial point for the rest of the article, and it might be good to make this point clearer.

Reply:
Good idea. I added the link.